lib/rom/plugins/relation/schema.rb

```diff
@@ -0,0 +1,52 @@
+require 'dry/data'
+
+module ROM
+  module Plugins
+    module Relation
+      module Schema
+        def self.included(klass)
+          super
+
+          klass.class_eval do
+            def self.schema(&block)
+              @__schema__ = Schema.create(&block) if block_given?
+              @__schema__
+            end
+          end
+        end
+
+        class Schema
+          attr_reader :attributes
+          attr_reader :coercer
+
+          def self.create(&block)
+            new.tap { |schema| schema.instance_eval(&block) }
+          end
+
+          def initialize(attributes = {}, coercer = Dry::Data::DSL.new(Dry::Data::Registry.new))
+            @attributes = attributes
+            @coercer = coercer
+          end
+
+          def attribute_names
+            attributes.keys
+          end
+
+          def attribute(name, type)
+            attributes[name] = coercer[type.to_s]
+          end
+
+          def apply(attributes = {})
+            self.attributes.each_with_object({}) do |(key, coercer), hash|
+              hash[key] = coercer.call(attributes[key])
+            end
+          end
+        end
```

Comment:
So, actually, I think we want a type-safe hash that we could build from the schema definition rather than re-using pieces of `dry-data` like that. First is to add that to `dry-data`, it should be cool since `Dry::Data::Struct` could just re-use it internally.